rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java

```diff
@@ -46,113 +74,496 @@
+                logger.info("Local jar " + file + " is obsolete or not present");
             } else {
-                System.out.println("Local jar " + file + " is up to date");
+                logger.info("Local jar " + file + " is up to date");
                 return true;
             }
 
         } catch (IOException e) {
-            e.printStackTrace();
+            logError("Error when contacting the remote url " + url, e);
         }
 
         return false;
     }
 
-    private static boolean makeNodeUpToDate() {
-        if (System.getProperty(NODE_URL_PROPERTY) != null) {
+    private static void logError(String message, Exception e) {
+        if (logger.isDebugEnabled()) {
+            logger.error(message, e);
+        } else {
+            logger.error(message + " : " + e.getMessage());
+        }
+    }
+
+    private long getRemoteLastModifiedMillis(String url) {
+        try {
+            URLConnection urlConnection = new URL(url).openConnection();
+            return urlConnection.getLastModified();
+        } catch (IOException e) {
+            logError("Error when contacting the remote url " + url, e);
+            return 0;
+        }
+    }
+
+    private boolean makeNodeUpToDate() {
+        if (!isLocalJarUpToDate(nodeJarUrl, nodeJarSaveAs)) {
+            logger.info("Downloading node.jar from " + nodeJarUrl + " to " + nodeJarSaveAs);
+
+            try {
+                File destination = new File(nodeJarSaveAs);
+                File lockFile = null;
+                FileLock lock = null;
 
-            String jarUrl = System.getProperty(NODE_URL_PROPERTY);
-            String jarFile = SCHEDULER_NODE_JAR;
+                if (destination.exists()) {
 
-            if (System.getProperty(NODE_JAR_PROPERTY) != null) {
-                jarFile = System.getProperty(NODE_JAR_PROPERTY);
+                    lockFile = new File(StandardSystemProperty.JAVA_IO_TMPDIR.value(), "lock");
+                    if (!lockFile.exists()) {
+                        lockFile.createNewFile();
+                    }
+
+                    logger.info("Getting the lock on " + lockFile.getAbsoluteFile());
+                    FileChannel channel = new RandomAccessFile(lockFile, "rw").getChannel();
+                    lock = channel.lock();
+
+                    if (isLocalJarUpToDate(nodeJarUrl, nodeJarSaveAs)) {
+                        logger.warn("Another process downloaded node.jar - don't do it anymore");
+                        logger.info("Releasing the lock on " + lockFile.getAbsoluteFile());
+                        lock.release();
+                        channel.close();
+
+                        return false;
+                    }
+                }
+                fetchUrl(nodeJarUrl, destination);
+                // Align the local file modification time with the remote url.
+                destination.setLastModified(getRemoteLastModifiedMillis(nodeJarUrl));
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Do something with the "boolean" value returned by "setLastModified". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS899)